model: Inform user if message was not sent successfully.

amanagr opened the pull request.

amanagr committed:
writebox: Inform user if message is successfully sent via footer.

sumanthvrao commented:
This looks like a nice addition! Just tried this and the 3 seconds seems fine for now, though we can probably reduce it a bit (based on more feedback of-course)

neiljp reviewed (commented):
@amanagr I like that the user will get more feedback here :)

However, right now there are a number of issues, including at least:
* when there is (briefly) no network, the UI locks up until it times-out (in `self.model.send*`?) before displaying the final message in the footer; the "Sending..." doesn't show.
* if the network outage is short enough, it *does* send the message, once connection is re-established.

The first issue is likely fairly easily resolved, but the second may be more complex?

The first commit is from another PR of yours?

amanagr commented:
Actually this was wip, I forgot to mention, Yeah the first commit is from another PR.

amanagr committed:
send_message: Combine functions for sending message.

Display "edited" if message is edited and "sent" if it is sent in footer.

neiljp commented:
I won't review until you're ready, but whether in this PR or a follow-up, we could also have something similar which managed other actions too; this was just made clear when czo was updated this afternoon :)

amanagr commented:
@neiljp I feel this is mostly ready. For the short network outrage issue, do we really need to deal with it since the message is being successfully sent in the end?

neiljp commented:
> @neiljp I feel this is mostly ready. For the short network outrage issue, do we really need to deal with it since the message is being successfully sent in the end?

To clarify, the issue here is that the message *was* sent, but the notice given to the user in the footer is that it *wasn't*.

amanagr commented:
@neiljp thanks for the find. I was able to reproduce. Seems like a tricky case. We send the request to client and client returns a response, that should end the thread right since we are not longpolling here or not doing any error_retries?

neiljp reviewed (commented):
@amanagr Other than the rebase needed after I merged the common commit from this PR and another, it seems like the last commit contains two changes?

I'm a little wary of having the model access the view via the controller, which wasn't the case when this occurred in the keypress method?

The lock could resolve the timing issue, though I'll wait until it's rebased and split to review further.

amanagr commented:
re: `last commit contains two changes?` -> Do you want to split the send_lock part?
re: `I'm a little wary of having the model access the view via the controller, which wasn't the case when this occurred in the keypress method?` -> We can add a method for updating footer to the controller so that it's easy for various functions in model (even in future) to update the footer.
re:`The lock could resolve the timing issue` -> Yeah it's a tricky case would love to have additional ideas.

amanagr commented on zulipterminal/model.py:
```diff
@@ -263,26 +265,24 @@ def mark_message_ids_as_read(self, id_list: List[int]) -> None:
+    def send_message(self, request: Dict[str, Any]) -> None:
+        if self.sending_message:
+            return
+        self.send_lock.acquire()
```
Okay!

neiljp commented:
* Re 2 changes, it felt like you had restructured and added the lock in the same commit? If the code stays in that form, it felt like the restructure should be in a separate commit, or just be part of the first one
* I understand that you want to place the lock in the model; it feels like it could also work with the lock in a controller method, and then that method could update the UI?

amanagr commented:
I remove the lock and asynch, it works fine without them both.

amanagr committed:
core: Add method to set_footer_text in controller.

This will allow setting footer directly via controller and avoid calls
to view.

neiljp commented:
@amanagr I'm observing the same behavior here as before; trying to send with no network:
* doesn't update the UI with 'Sending...' at all
* pauses the UI for the non-longpolling 15s
* shows the the failure message
* if reconnected "quick enough", does actually send the message?

I expect we want the 15s timeout for actually sending, which suggests we do want something with asynch?

With network, it's fine, in that it updates the bar with a temporary success message, but that's not the complex part.

amanagr committed:
model: Make send_message async and block multiple calls to it.

sumanthvrao commented:
@amanagr A small extension here would be to have the footer display when a message is sent out the current narrow (somewhat similar to the web app message we get).

neiljp commented:
Another extension could include to indicate that the message is below the current view, though that might be difficult to determine, I'm not sure?

amanagr commented:
Thanks for the suggestions!

amanagr commented:
@neiljp I am unable to figure out why the message is being successfully sent after we get an error response from the server in this case. Do you have any ideas as to why this might be happening given your history with the python-zulip-api?
